src/core/qgsvectorlayerfeatureiterator.cpp

```diff
@@ -904,6 +905,42 @@ bool QgsVectorLayerFeatureIterator::checkGeometryValidity( const QgsFeature &fea
         return false;
       }
       break;
+
+    case QgsFeatureRequest::GeometryFixInvalid:
+      if ( !feature.geometry().isGeosValid() )
+      {
+        QgsGeometry outputGeometry = feature.geometry().makeValid();
+        if ( outputGeometry.isNull() )
+        {
+          return  false;
+        }
+
+        if ( outputGeometry.wkbType() == QgsWkbTypes::Unknown ||
+             QgsWkbTypes::flatType( outputGeometry.wkbType() ) == QgsWkbTypes::GeometryCollection )
+        {
+          // keep only the parts of the geometry collection with correct type
+          const QVector< QgsGeometry > tmpGeometries = outputGeometry.asGeometryCollection();
+          QVector< QgsGeometry > matchingParts;
+          for ( const QgsGeometry &g : tmpGeometries )
+          {
+            if ( g.type() == feature.geometry().type() )
+              matchingParts << g;
+          }
+          if ( !matchingParts.empty() )
+            outputGeometry = QgsGeometry::collectGeometry( matchingParts );
+          else
+            outputGeometry = QgsGeometry();
+        }
+
+        outputGeometry.convertToMultiType();
```

Comment:
```suggestion
```

That's not a safe move -- I'd only do this if the original geometry was also multi type.

Reply:
To be honest in my manual tests this is the part that actually fixed things, especially for geometries that cross themselves. I feel like removing this greatly reduce the fixing capabilities. I'd be fine with a safer approach but would greatly suggest keeping this function in some form.

Reply:
To be discussed with flags.

Reply:
done as suggested.